### PR TITLE
Fix doubled-find_package()

### PR DESCRIPTION
**Pull request by MoriKen254**
Fixed https://github.com/ros-controls/ros_controllers/issues/378

for controllers as follows

- joint_trajectory_controller
- diff_drive_controller
- four_wheel_steering_controller
- ackermann_steering_controller

### CODE REVIEW

**Review by mathias-luedtke — CHANGES_REQUESTED** [review]
Thanks for the effort, but that's not what I meant.
The test dependencies should not get mixed in.

There are two ways:
1. Resolve test dependencies individually (this gets tricky with more dependencies)
2. Resolve ALL test dependencies in a separate `find_package` call (overwrites the first)

I am not sure which is the best way to go.

**Comment by MoriKen254** [comment]
Thank you for your comment and excuse me for my misunderstanding.

Could https://github.com/ros-controls/ros_controllers/pull/380/commits/2d4bab0cf2b5a64ad4077a56ff8d1e850e4f4bd1 be a solution? I overwrote all dependent pkgs in test section.

Just in my opinion, in terms of maintainability it sounds better to do this even though it may look redundant. Checking dependency in test again and removing pkgs unrelated only to test afterwards sounds like a complicated and tough task...

At least, wouldn't this be better compared to the previous one?

File: ros_controllers/package.xml [hunk 1]
```diff
@@ -23,6 +23,7 @@
   <exec_depend>effort_controllers</exec_depend>
   <exec_depend>force_torque_sensor_controller</exec_depend>
   <exec_depend>forward_command_controller</exec_depend>
+  <exec_depend>four_wheel_steering_controller</exec_depend>
```
Comment (mathias-luedtke) [inline ros_controllers/package.xml:26]: This decision is up to @bmagyar

File: joint_trajectory_controller/CMakeLists.txt [hunk 2]
```diff
@@ -48,17 +51,30 @@ add_library(${PROJECT_NAME} src/joint_trajectory_controller.cpp
...
+      roscpp
+      rostest
+      rosunit
+      rqt_plot
+      trajectory_msgs
+      urdf
       xacro
   )
 
-  find_package(rostest REQUIRED)
```
Comment (mathias-luedtke) [inline joint_trajectory_controller/CMakeLists.txt:61]: I would prefer to resolve `rostest` separately. (in all packages)

File: ackermann_steering_controller/package.xml [hunk 3]
```diff
@@ -32,8 +32,9 @@
   <test_depend>controller_manager</test_depend>
   <test_depend>gazebo_ros</test_depend>
   <test_depend>geometry_msgs</test_depend>
-  <test_depend>rosunit</test_depend>
   <test_depend>rostest</test_depend>
+  <test_depend>rosunit</test_depend>
+  <test_depend>rviz</test_depend>
```
Comment (mathias-luedtke) [inline ackermann_steering_controller/package.xml:37]: I think it is better to not depend on rviz. It should not be crucial for the test.

File: ackermann_steering_controller/CMakeLists.txt [hunk 4]
```diff
@@ -49,13 +49,23 @@ install(FILES ackermann_steering_controller_plugins.xml
 
 if (CATKIN_ENABLE_TESTING)
   find_package(catkin REQUIRED COMPONENTS
+    controller_interface
```
Comment (mathias-luedtke) [inline ackermann_steering_controller/CMakeLists.txt]: only compile-time dependencies should be listed here

**Review by mathias-luedtke — CHANGES_REQUESTED** [review]
Yes, this looks better.
But I am still unsure about which way to go here..
However, it makes sense to check/review the dependencies, even if we decide to go the other way later.

File: ros_controllers/package.xml (hunk 1 above)
Comment (bmagyar) [reply ros_controllers/package.xml:26]: Yes this is intended. The controller is in the repo, let's go all the way.

File: ackermann_steering_controller/package.xml (hunk 3 above)
Comment (bmagyar) [reply ackermann_steering_controller/package.xml:37]: And they should already be behind a `show_rviz` or `gui` flag.

File: joint_trajectory_controller/CMakeLists.txt (hunk 2 above)
Comment (MoriKen254) [reply joint_trajectory_controller/CMakeLists.txt:61]: Fixed at https://github.com/ros-controls/ros_controllers/pull/380/commits/e745c484517543b00b25111f5aa9dfe749c236ed

File: ackermann_steering_controller/package.xml (hunk 3 above)
Comment (MoriKen254) [reply ackermann_steering_controller/package.xml:37]: removed rviz at https://github.com/ros-controls/ros_controllers/pull/380/commits/e745c484517543b00b25111f5aa9dfe749c236ed

File: ackermann_steering_controller/CMakeLists.txt (hunk 4 above)
Comment (MoriKen254) [reply ackermann_steering_controller/CMakeLists.txt]: Tried to fix at https://github.com/ros-controls/ros_controllers/pull/380/commits/e745c484517543b00b25111f5aa9dfe749c236ed looking at test codes inside

**Comment by MoriKen254** [comment]
Thanks for your generous comments. I got your point. 

hope this PR will be a kinda current milestone to help you review dependencies and find something to decide which way to go.

**Comment by bmagyar** [comment]
@ipa-mdl Should we rebase & fix this one?

**Comment by bmagyar** [comment]
I've rebased and resolved the conflict on this one, not sure if all changes are required though. 

@matthew-reynolds care to take over this one?

**Comment by matthew-reynolds** [comment]
Sure thing. I'll base it off what we learned/decided in https://github.com/ros-controls/ros_control/pull/404

**Comment by bmagyar** [comment]
This should all be clean in Noetic

**Pull request closed without merging**